src/Interpreters/Cache/FileCache.cpp

```diff
@@ -946,6 +955,116 @@ bool FileCache::tryReserve(
+    EvictionCandidates eviction_candidates;
+
+    bool limits_satisfied = true;
+    try
+    {
+        /// Collect at most `keep_up_free_space_remove_batch` elements to evict,
+        /// (we use batches to make sure we do not block cache for too long,
+        /// by default the batch size is quite small).
+        limits_satisfied = main_priority->collectCandidatesForEviction(
+            desired_size, desired_elements_num, keep_up_free_space_remove_batch, stat, eviction_candidates, lock);
+
+#ifdef ABORT_ON_LOGICAL_ERROR
+        /// Let's make sure that we correctly processed the limits.
+        if (limits_satisfied && eviction_candidates.size() < keep_up_free_space_remove_batch)
+        {
+            const auto current_size = main_priority->getSize(lock);
+            chassert(current_size >= stat.total_stat.releasable_size);
+            chassert(!size_limit
+                        || current_size <= desired_size
+                        || current_size - stat.total_stat.releasable_size <= desired_size);
+
+            const auto current_elements_count = main_priority->getElementsCount(lock);
+            chassert(current_elements_count >= stat.total_stat.releasable_count);
+            chassert(!elements_limit
+                        || current_elements_count <= desired_elements_num
+                        || current_elements_count - stat.total_stat.releasable_count <= desired_elements_num);
+        }
+#endif
+
+        if (shutdown)
+            return;
+
+        if (eviction_candidates.size() > 0)
+        {
+            LOG_TRACE(log, "Current usage {}/{} in size, {}/{} in elements count "
+                    "(trying to keep size ration at {} and elements ratio at {}). "
+                    "Collected {} eviction candidates, "
+                    "skipped {} candidates while iterating",
+                    main_priority->getSize(lock), size_limit,
+                    main_priority->getElementsCount(lock), elements_limit,
+                    desired_size, desired_elements_num,
+                    eviction_candidates.size(), stat.total_stat.non_releasable_count);
+
+            lock.unlock();
+
+            /// Remove files from filesystem.
+            eviction_candidates.evict();
+
+            /// Take lock again to finalize eviction,
+            /// e.g. to update the in-memory state.
+            lock.lock();
+            eviction_candidates.finalize(nullptr, lock);
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+
+        if (eviction_candidates.size() > 0)
+            eviction_candidates.finalize(nullptr, lockCache());
+
+        /// Let's catch such cases in ci,
+        /// in general there should not be exceptions.
+        chassert(false);
+    }
+
+    LOG_TRACE(log, "Free space ratio keeping thread finished in {} ms", watch.elapsedMilliseconds());
+
+    if (limits_satisfied)
+        keep_up_free_space_ratio_task->scheduleAfter(general_reschedule_ms);
```

Comment:
Maybe a check that the task was scheduled successfully (both `scheduleAfter` and `schedule` return `bool`).